### PR TITLE
Use nbctl transactions for node and gateway setup operations

### PR DESCRIPTION
**Pull request by dcbw**
Create a simple Go struct to batch up ovn-nbctl commands and run them all in the same transaction. Then use that in node setup paths.

### CODE REVIEW

**Comment by dcbw** [comment]
@ovn-org/ovn-kubernetes-committers

**Comment by dcbw** [comment]
CI failures are Kind setup errors due to Fedora mirrors being out of sync with repodata.

**Comment by trozet** [comment]
@dcbw is this necessary if we are moving to ebay IDL?

**Comment by dcbw** [comment]
> @dcbw is this necessary if we are moving to ebay IDL?

@trozet not strictly, but it's orthogonal to https://github.com/ovn-org/ovn-kubernetes/pull/1335 until we extend the bindings to cover node setup. But this PR makes it easier for bindings since now we know exactly what stuff can be batched into the same transaction.

**Comment by dcbw** [comment]
@girishmg what do you think?

**Comment by dcbw** [comment]
@trozet @girishmg PTAL thanks!

File: go-controller/pkg/util/ovs.go [hunk 1]
```diff
@@ -432,6 +432,38 @@ func RunOVNNbctl(args ...string) (string, string, error) {
...
+// NBTxn hold parts of an ovn-nbctl transaction request
+type NBTxn struct {
+	args    []string
+	txnArgs []string
+	env     []string
+}
+
+// NewNBTxn returns a new ovn-nbctl transaction request object
+func NewNBTxn() *NBTxn {
+	args, env := getNbctlArgsAndEnv(ovsCommandTimeout, []string{}...)
```
Comment (aojea) [inline go-controller/pkg/util/ovs.go:444]: is guaranteed that it always return the right `args` and `env` parameters?
or should it fail to create the new object?

Comment (dcbw) [reply go-controller/pkg/util/ovs.go:444]: Yeah, that function has no way to return an error so I think we are OK.

**Comment by dcbw** [comment]
@girishmg @danwinship @trozet how does this look now?

**Comment by danwinship** [comment]
I guess? Seems fine (though also, seems like it could have waited until after go-ovn).

It annoys me that the command name isn't always the first argument to `txn.Add()` but I guess that's an `ovn-nbctl` syntax problem...

**Comment by dcbw** [comment]
> I guess? Seems fine (though also, seems like it could have waited until after go-ovn).

We'd need to do the same thing for go-ovn (eg consolidating separate commands) anyway, might as well do it now. Also this doesn't conflict with the go-ovn bindings work in https://github.com/ovn-org/ovn-kubernetes/pull/1335 since that's focused on pods, not node setup.

**Comment by danwinship** [comment]
@dcbw needs rebase

**Comment by trozet** [comment]
I would personally rather just wait and do things in go-ovn.

**Pull request closed without merging**